util/math_reference.h

```diff
@@ -305,21 +305,45 @@ sycl_cts::resultRef<sycl::marray<T, N>> abs(sycl::marray<T, N> a) {
 
 /* absolute difference */
 template <typename T>
-T abs_diff(T a, T b) {
-  T h = (a > b) ? a : b;
-  T l = (a <= b) ? a : b;
-  return h - l;
+sycl_cts::resultRef<T> abs_diff(T a, T b) {
+  if constexpr (std::is_unsigned_v<T>) {
+    T h = (a > b) ? a : b;
+    T l = (a <= b) ? a : b;
+    return h - l;
+  } else {
+    using U = std::make_unsigned_t<T>;
+    constexpr T TMax = std::numeric_limits<T>::max();
+    // For unsigned integers, negative values might not be representable by
+    // an absolute value, so we compute the overflow separately.
+    T ao = a < -TMax ? a + TMax : 0;
+    T bo = b < -TMax ? b + TMax : 0;
+    // Then find the absolute values (without the overflow) and compute the
+    // unsigned difference.
+    U au = static_cast<U>(std::abs(a - ao));
+    U bu = static_cast<U>(std::abs(b - bo));
+    U h = (au > bu) ? au : bu;
+    U l = (au <= bu) ? au : bu;
+    U result = (a < 0) == (b < 0) ? h - l : h + l;
+    // Now re-add the absolute distance in overflow. Both ao and bo will be
+    // <= 0.
+    result += (ao <= bo ? U(bo - ao) : U(ao - bo));
+    // If the unsigned difference is larger than the signed maximum value, we
+    // have UB. Otherwise we have our valid result.
+    return result > U(TMax) ? sycl_cts::resultRef<T>(0, true) : T(result);
+  }
```

Comment:
```suggestion
  using U = std::make_unsigned_t<T>;
  auto normalize = [](T& x) {
    constexpr U TMin = static_cast<U>(std::numeric_limits<T>::min());
    U xu = static_cast<U>(x);
    if (x < 0)
      xu -= TMin;
    else
      xu += TMin;
    return xu;
  };
  U au = normalize(a);
  U bu = normalize(b);
  U h = (au > bu) ? au : bu;
  U l = (au <= bu) ? au : bu;
  U result = h - l;
  return result > static_cast<U>(std::numeric_limits<T>::max())
             ? sycl_cts::resultRef<T>(0, true)
             : T(result);
```

Suggestion based on https://github.com/KhronosGroup/SYCL-CTS/pull/842#issuecomment-1856165033 to illustrate how it could be integrated here.

Reply:
I was thinking to https://godbolt.org/z/9Ma6nvGsz but I think I can even optimizing it further.

Reply:
Doesn't that solution rely on underflowing behavior if either `a` or `b` is negative? AFAIK that is UB.

Reply:
The shorter simplified version: https://godbolt.org/z/4Wb8x6GKG

Reply:
```suggestion
  using U = std::make_unsigned_t<T>;
  T h = (a > b) ? a : b;
  T l = (a <= b) ? a : b;
  // Use knowledge of 2-complement encoding while we know the difference is
  // positive, so no overflow/underflow.
  U result = static_cast<U>(h) - static_cast<U>(l);
  return result > static_cast<U>(std::numeric_limits<T>::max())
             ? sycl_cts::resultRef<T>(0, true)
             : T(result);
```

However, I worry that the comment is misleading. We are potentially causing an underflow if one of the operands are negative. Say for example `a = 1000` and `b = -1000`, the the static cast to unsigned for these would result in the computation `1000 - 4294966296`. That said, it is now for unsigned integer types, which I believe has well-defined overflow/underflow behavior (modulo 2^n).

Reply:
I have applied your suggestion since it should be fine. If you agree that the comment is slightly misleading we can change it.

Reply:
Yes, by "so no overflow/underflow." I mean with `signed`.
Once we are back in the `unsigned` world we are good, since as you said we are just doing modulo 2^n arithmetic.
I let you tweak the comment to explain this in a better way. :-)

Reply:
The cast is redundant in
```
return result > static_cast<U>(std::numeric_limits<T>::max())
```
since in C/C++ there is this implicit (confusing...) conversion of signed to unsigned.

Reply:
I have updated the comment and removed the redundant cast. 👍